blob/conformance.go

```diff
@@ -0,0 +1,460 @@
+		if err == nil {
+			err = w.Close()
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer func() {
+			_ = b.Delete(ctx, key)
+		}()
+
+		for _, tc := range tests {
+			t.Run(tc.name, func(t *testing.T) {
+				b, done := makeBkt(t)
+				defer done()
+
+				r, err := b.NewRangeReader(ctx, tc.key, tc.offset, tc.length)
+				if (err != nil) != tc.wantErr {
+					t.Errorf("got err %v want error %v", err, tc.wantErr)
+				}
+				if err != nil {
+					return
+				}
+				defer r.Close()
+				// Make the buffer bigger than needed to make sure we actually only read
+				// the expected number of bytes.
+				got := make([]byte, tc.wantReadSize+10)
+				n, err := r.Read(got)
+				if err != nil {
+					t.Errorf("unexpected error during read: %v", err)
+				}
+				if int64(n) != tc.wantReadSize {
+					t.Errorf("got read length %d want %d", n, tc.wantReadSize)
+				}
+				if !cmp.Equal(got[:tc.wantReadSize], tc.want) {
+					t.Errorf("got %q want %q", string(got), string(tc.want))
+				}
+				// TODO(issue #305): Fails for S3.
+				/*
+					if r.Size() != contentSize {
+						t.Errorf("got size %d want %d", r.Size(), contentSize)
+					}
+				*/
+				r.Close()
+			})
+		}
+	})
+}
+
+// testAttributes tests the behavior of attributes returned by Reader.
+func testAttributes(t *testing.T, makeBkt bucketMaker) {
+	const (
+		key         = "blob-for-attributes"
+		contentType = "text/plain"
+	)
+	content := []byte("Hello World!")
+
+	ctx := context.Background()
+	t.Run("Attributes", func(t *testing.T) {
+
+		// Create a blob to be read by sub-tests below.
+		b, done := makeBkt(t)
+		defer done()
+
+		opts := &WriterOptions{
+			ContentType: contentType,
+		}
+		w, err := b.NewWriter(ctx, key, opts)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = w.Write(content)
+		if err == nil {
+			err = w.Close()
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer func() {
+			_ = b.Delete(ctx, key)
+		}()
+
+		t.Run("ContentType", func(t *testing.T) {
+			r, err := b.NewRangeReader(ctx, key, 0, 0)
+			if err != nil {
+				t.Fatalf("failed NewRangeReader: %v", err)
+			}
+			defer r.Close()
+			if r.ContentType() != contentType {
+				t.Errorf("got ContentType %q want %q", r.ContentType(), contentType)
+			}
+		})
+
+		// TODO(issue #303): Fails for GCS.
+		/*
+			t.Run("Size", func(t *testing.T) {
+				r, err := b.NewRangeReader(ctx, key, 0, 0)
+				if err != nil {
+					t.Errorf("failed NewRangeReader: %v", err)
+				}
+				defer r.Close()
+				if r.Size() != int64(len(content)) {
+					t.Errorf("got Size %d want %d", r.Size(), len(content))
+				}
+			})
+		*/
+
+		t.Run("ModTime", func(t *testing.T) {
+			r, err := b.NewRangeReader(ctx, key, 0, 0)
+			if err != nil {
+				t.Fatalf("failed NewRangeReader: %v", err)
+			}
+			defer r.Close()
+			t1 := r.ModTime()
+			if t1.IsZero() {
+				// This provider doesn't support ModTime.
+				return
+			}
+			// Touch the file after a couple of seconds and make sure ModTime changes.
+			time.Sleep(2 * time.Second)
```

Comment:
Having a sleep in a test makes me uncomfortable, but I don't have a concrete recommendation of what should be done instead.

Reply:
Agree. At first I thought the sleep could be done in --record mode and avoided in replay. It can, for provider implementations that use record/replay, but e.g. fileblob breaks without the sleep. I don't think there's a feasible way to mock or stub time at the provider.

There is probably a way to identify replay tests and avoid this Sleep for them specifically.